### PR TITLE
[CMake] Replace early swift-syntax with FetchContent

### PR DESCRIPTION
**Pull request by rintaro**
Use FetchContent to include swift-syntax directly in swift. This can be thought of as an add_subdirectory for a directory outside the root.

The default build directory will be _deps/swiftsyntax-subbuild/, though the modules and shared libraries will be built in lib/swift/host by passing down SWIFT_HOST_LIBRARIES_DEST_DIR to avoid copying them as we were doing previously.

### CODE REVIEW

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

**Comment by finagolfin** [comment]
Spurred by #68401, I started looking into cross-compiling swift-syntax for the native Android compiler, only to find early-swift-syntax doesn't support cross-compilation. With a tweak or two to this pull, I think it can be made to work: is that the plan?

Because I don't know about macOS, but the recent shipping of swift-syntax with the linux Swift compiler means that cross-compiling the toolchain is currently broken on Unix platforms, ie swift-syntax won't be cross-compiled.

Also, any plans to backport this pull to 5.9?

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
@finagolfin 
> Spurred by https://github.com/apple/swift/pull/68401, I started looking into cross-compiling swift-syntax for the native Android compiler, only to find early-swift-syntax doesn't support cross-compilation. With a tweak or two to this pull, I think it can be made to work: is that the plan?

The _main_ motivation of this PR is that single `cmake --build /path/to/swift-build-dir`  invocation rebuilds swift-syntax libraries if modified. But I'm aware of that the `earlyswiftsyntax` didn't support cross compiling, and this PR may make it easier. I'd appreciate it if you can check (and fix) it after we merge this PR.

> Also, any plans to backport this pull to 5.9?

I cannot promise anything, but I personally want this in 5.9 too.

File: cmake/modules/AddPureSwift.cmake [hunk 1]
```diff
@@ -46,20 +46,7 @@ function(_add_host_swift_compile_options name)
...
-
-  if(SWIFT_HOST_VARIANT_SDK STREQUAL "ANDROID")
-    set(DEPLOYMENT_VERSION ${SWIFT_ANDROID_API_LEVEL})
-  endif()
-
-  get_target_triple(target target_variant "${SWIFT_HOST_VARIANT_SDK}" "${SWIFT_HOST_VARIANT_ARCH}"
-    MACCATALYST_BUILD_FLAVOR ""
-    DEPLOYMENT_VERSION "${DEPLOYMENT_VERSION}")
-
+  swift_get_host_triple(target)
```
Comment (bnbarham) [inline cmake/modules/AddPureSwift.cmake:49]: Can this just use `SWIFT_HOST_TRIPLE`?

**Comment by finagolfin** [comment]
> But I'm aware of that the earlyswiftsyntax didn't support cross compiling, and this PR may make it easier. I'd appreciate it if you can check (and fix) it after we merge this PR.

I got the flags added locally to cross-compile this and the regex parser, will submit that once this is in.

I've also backported this to 5.9, along with #68041. Since that adds more checks of `SWIFT_SWIFT_PARSER`, whichever one of these two pulls goes in first will mean the second will have to be updated.

File: cmake/modules/AddPureSwift.cmake (hunk 1 above)
Comment (rintaro) [reply cmake/modules/AddPureSwift.cmake]: Updated

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please test

**Review by bnbarham — APPROVED** [review]
🥳

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

**Pull request merged**